File: backend/src/models/Session.js
const mongoose = require('mongoose');

const { Schema } = mongoose;

const sessionSchema = new Schema(
  {
    user: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
    },
    refreshTokenHash: {
      type: String,
      required: true,
      index: true,
      unique: true,
    },
    expiresAt: {
      type: Date,
      required: true,
    },
    revokedAt: {
      type: Date,
      default: null,
    },
    userAgent: {
      type: String,
      maxlength: 512,
    },
    ipAddress: {
      type: String,
      maxlength: 64,
    },
    metadata: {
      type: Schema.Types.Mixed,
    },
  },
  {
    timestamps: true,
  }
);

sessionSchema.index(
  { expiresAt: 1 },
  {
    expireAfterSeconds: 0,
    partialFilterExpression: { revokedAt: null },
  }
);

module.exports = mongoose.model('Session', sessionSchema);

File: backend/src/models/Session.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from beanie import Indexed
from beanie.odm.fields import PydanticObjectId
from pymongo import IndexModel

from .base import TimeStampedDocument


class Session(TimeStampedDocument):
    user: Indexed(PydanticObjectId)  # type: ignore[assignment]
    refreshTokenHash: Indexed(str, unique=True)  # type: ignore[assignment]
    expiresAt: datetime
    revokedAt: Optional[datetime] = None
    userAgent: Optional[str] = None
    ipAddress: Optional[str] = None
    metadata: Optional[dict] = None

    class Settings:
        name = "sessions"
        use_revision = False
        indexes = [
            IndexModel(
                [("expiresAt", 1)],
                expireAfterSeconds=0,
                partialFilterExpression={"revokedAt": None},
            )
        ]

